docs/design/2018-08-29-new-planner.md

```diff
@@ -0,0 +1,411 @@
+the pseudo code:
+
+```go
+func implGroup(g *Group, reqPhysProp *PhysicalProperty, costLimit float64) (groupImpl PhysicalPlan) {
+	if g.implemented(reqPhysProp) {
+		return g.getImplementation(reqPhysProp)
+	}
+
+	// handle implementation rules for each equivalent expression.
+    ...
+
+	// handle enforcing rules for the required physical property.
+    ...
+
+	g.insertImpl(reqPhysProp, groupImpl)
+	return groupImpl
+}
+```
+
+In order to find the best physical implementation for the group, we need to
+enumerate all the possible implementations for each group expression under the
+required physical property, the procedure to handle the implementation rules for
+each equivalent expression is:
+
+```go
+// handle implementation rules for each equivalent expression.
+for _, curExpr := range g.equivalents {
+	for _, impl := range implGroupExpr(curExpr) {
+		impl.setCumCost(impl.getSelfCost())
+		for i, childGroup := range curExpr.children {
+			childImpl := implGroup(childGroup, impl.getChildReqProp(i), costLimit-impl.getCumCost())
+			if childImpl == nil {
+				cumCost = math.MaxFloat64
+				break
+			}
+			impl.setChild(i, childImpl)
+			impl.setCumCost(impl.getCumCost() + childImpl.getCumCost())
+		}
+		if groupImpl.getCumCost() > impl.getCumCost() {
+			groupImpl = impl
+		}
+	}
+}
+```
+
+To enumerate all the implementation for an expression, we have to enumerate all
+the applicable implementation rules on that expression, and calculate the
+self-cost for physical implementation:
+
+```go
+func implGroupExpr(cur *GroupExpr, reqPhysProp *PhysicalProperty) (impls []PhysicalPlan) {
+	for _, rule := range GetImplementationRules(cur.exprNode) {
+		if !rule.match(cur, reqPhysProp) {
+			continue
+		}
+		impl := rule.onImplement(cur)
+		impl.calcSelfCost(cur.children)
+		impls = append(impls, impl)
+	}
+	return impls
+}
+```
+
+Also, the enforcing rules should be considered:
+
+```go
+// handle enforcing rules for the required physical property.
+for _, rule := range getEnforcerRules(reqPhysProp) {
+	impl, newReqPhysProp := rule.onEnforce(reqPhysProp)
+	impl.calcSelfCost(g)
+	impl.setCumCost(impl.getSelfCost())
+	childImpl := implGroup(g, newReqPhysProp, costLimit-impl.getCumCost())
+	impl.setChild(0, childImpl)
+	impl.setCumCost(impl.getCumCost() + childImpl.getCumCost())
+	if groupImpl.getCumCost() > impl.getCumCost() {
+		groupImpl = impl
+	}
+}
+```
+
+## Rationale
+
+## Compatibility
+
+No
+
+## Implementation
+
+The first step to implement the new planner is to add a session variable named
+`tidb_enable_volcano_planner` to control whether to use the new planner. Once
+this variable is set, all the optimization steps are handed to the new planner.
+The procedure of converting the abstract syntax tree to a logical plan is
+remained unchanged. The constructed physical plan is also compatible to the
+existing physical plan. The only affection should be the optimization algorithm.
```

Comment:
The only affection should be the optimization algorithm.
-> It only affects the optimization algorithm.
     
Note: Generally, "affection" means "love" or "fondness".